content/03_CreatingObservableSequences.md

```diff
@@ -0,0 +1,1274 @@
+
+```cs
+IObservable<string> source = Observable.FromAsync(() => Task.Run(() =>
+{
+    Console.WriteLine("Task running...");
+    return "Test";
+}));
+```
+
+Subscribing twice to this produces slightly different output:
+
+```
+Task running...
+Task running...
+Test
+Test
+completed
+completed
+```
+
+Notice that this executes the task twice, once for each call to `Subscribe`. `FromAsync` can do this because instead of passing a `Task<T>` we pass a callback that returns a `Task<T>`. It calls that when we call `Subscribe`, so each subscriber essentially gets their own task.
+
+If I want to use `async` and `await` to define my task, then I don't need to bother with the `Task.Run` because an `async` lambda creates a `Func<Task<T>>`, which is exactly the type `FromAsync` wants:
+
+```cs
+IObservable<string> source = Observable.FromAsync(async () =>
+{
+    Console.WriteLine("Task running...");
+    await Task.Delay(50);
+    return "Test";
+});
+```
+
+There is a subtle difference with this though. When I used `Task.Run` the lambda ran on a task pool thread from the start. But when I write it this way, the lambda will begin to run on whatever thread calls `Subscribe`. It's only when it hits the first `await` that it returns (and the call to `Subscribe` will then return), with the remainder of the method running on the thread pool.
+
+### From IEnumerable&lt;T&gt;
+
+Rx defines another extension method called `ToObservable`, this time for `IEnumerable<T>`. In earlier chapters I described how `IObservable<T>` was designed to represent the same basic abstraction as `IEnumerable<T>`, with the only difference being the mechanism we use to obtain the elements in the sequence: with `IEnumerable<T>`, we write code that _pulls_ values out of the collection (e.g., a `foreach` loop), whereas `IObservable<T>` _pushes_ values to us by invoking `OnNext` on our `IObserver<T>`.
+
+We could write code that bridges from _pull_ to _push_:
+
+```csharp
+// Example code only - do not use!
+public static IObservable<T> ToObservableOversimplified<T>(this IEnumerable<T> source)
+{
+    return Observable.Create<T>(o =>
+    {
+        foreach (var item in source)
+        {
+            o.OnNext(item);
+        }
+
+        o.OnComplete();
+
+        // Incorrectly ignoring unsubscription.
+        return Disposable.Empty;
+    });
+}
+```
+
+This crude implementation conveys the basic idea, but it is naive. It does not attempt to handle unsubscription, and it's not easy to fix that when using `Observable.Create` for this particular scenario. And as we will see later in the book, Rx sources that might try to deliver large numbers of events in quick succession should integrated with Rx's concurrency model. The implementation that Rx supplies does of course cater for all of these tricky details. That makes it rather more complex, but that's Rx's problem; you can think of it as being logically equivalent to the code shown above, but without the shortcomings.
```

Comment:
This is a profound and repeated message. Always figure out how to compose Rx operators to do what you need because, in all probability, you will get it wrong in subtle and near-impossible-to-spot ways.